// @remix-run/node/dist/magicExports/remix.js
/**
 * @remix-run/node v1.7.2
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var node = require('@remix-run/node');
Object.defineProperty(exports, 'createCookie', {
  enumerable: true,
  get: function () { return node.createCookie; }
});
Object.defineProperty(exports, 'createCookieSessionStorage', {
  enumerable: true,
  get: function () { return node.createCookieSessionStorage; }
});
Object.defineProperty(exports, 'createFileSessionStorage', {
  enumerable: true,
  get: function () { return node.createFileSessionStorage; }
});
Object.defineProperty(exports, 'createMemorySessionStorage', {
  enumerable: true,
  get: function () { return node.createMemorySessionStorage; }
});
Object.defineProperty(exports, 'createSessionStorage', {
  enumerable: true,
  get: function () { return node.createSessionStorage; }
});
Object.defineProperty(exports, 'unstable_createFileUploadHandler', {
  enumerable: true,
  get: function () { return node.unstable_createFileUploadHandler; }
});
Object.defineProperty(exports, 'unstable_createMemoryUploadHandler', {
  enumerable: true,
  get: function () { return node.unstable_createMemoryUploadHandler; }
});
Object.defineProperty(exports, 'unstable_parseMultipartFormData', {
  enumerable: true,
  get: function () { return node.unstable_parseMultipartFormData; }
});
