// @remix-run/react/dist/magicExports/remix.js
/**
 * @remix-run/react v1.7.2
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var react = require('@remix-run/react');
Object.defineProperty(exports, 'Form', {
  enumerable: true,
  get: function () { return react.Form; }
});
Object.defineProperty(exports, 'Link', {
  enumerable: true,
  get: function () { return react.Link; }
});
Object.defineProperty(exports, 'Links', {
  enumerable: true,
  get: function () { return react.Links; }
});
Object.defineProperty(exports, 'LiveReload', {
  enumerable: true,
  get: function () { return react.LiveReload; }
});
Object.defineProperty(exports, 'Meta', {
  enumerable: true,
  get: function () { return react.Meta; }
});
Object.defineProperty(exports, 'NavLink', {
  enumerable: true,
  get: function () { return react.NavLink; }
});
Object.defineProperty(exports, 'PrefetchPageLinks', {
  enumerable: true,
  get: function () { return react.PrefetchPageLinks; }
});
Object.defineProperty(exports, 'RemixBrowser', {
  enumerable: true,
  get: function () { return react.RemixBrowser; }
});
Object.defineProperty(exports, 'RemixServer', {
  enumerable: true,
  get: function () { return react.RemixServer; }
});
Object.defineProperty(exports, 'Scripts', {
  enumerable: true,
  get: function () { return react.Scripts; }
});
Object.defineProperty(exports, 'ScrollRestoration', {
  enumerable: true,
  get: function () { return react.ScrollRestoration; }
});
Object.defineProperty(exports, 'useActionData', {
  enumerable: true,
  get: function () { return react.useActionData; }
});
Object.defineProperty(exports, 'useBeforeUnload', {
  enumerable: true,
  get: function () { return react.useBeforeUnload; }
});
Object.defineProperty(exports, 'useCatch', {
  enumerable: true,
  get: function () { return react.useCatch; }
});
Object.defineProperty(exports, 'useFetcher', {
  enumerable: true,
  get: function () { return react.useFetcher; }
});
Object.defineProperty(exports, 'useFetchers', {
  enumerable: true,
  get: function () { return react.useFetchers; }
});
Object.defineProperty(exports, 'useFormAction', {
  enumerable: true,
  get: function () { return react.useFormAction; }
});
Object.defineProperty(exports, 'useLoaderData', {
  enumerable: true,
  get: function () { return react.useLoaderData; }
});
Object.defineProperty(exports, 'useMatches', {
  enumerable: true,
  get: function () { return react.useMatches; }
});
Object.defineProperty(exports, 'useSubmit', {
  enumerable: true,
  get: function () { return react.useSubmit; }
});
Object.defineProperty(exports, 'useTransition', {
  enumerable: true,
  get: function () { return react.useTransition; }
});
Object.defineProperty(exports, 'Outlet', {
  enumerable: true,
  get: function () { return react.Outlet; }
});
Object.defineProperty(exports, 'useHref', {
  enumerable: true,
  get: function () { return react.useHref; }
});
Object.defineProperty(exports, 'useLocation', {
  enumerable: true,
  get: function () { return react.useLocation; }
});
Object.defineProperty(exports, 'useNavigate', {
  enumerable: true,
  get: function () { return react.useNavigate; }
});
Object.defineProperty(exports, 'useNavigationType', {
  enumerable: true,
  get: function () { return react.useNavigationType; }
});
Object.defineProperty(exports, 'useOutlet', {
  enumerable: true,
  get: function () { return react.useOutlet; }
});
Object.defineProperty(exports, 'useOutletContext', {
  enumerable: true,
  get: function () { return react.useOutletContext; }
});
Object.defineProperty(exports, 'useParams', {
  enumerable: true,
  get: function () { return react.useParams; }
});
Object.defineProperty(exports, 'useResolvedPath', {
  enumerable: true,
  get: function () { return react.useResolvedPath; }
});
Object.defineProperty(exports, 'useSearchParams', {
  enumerable: true,
  get: function () { return react.useSearchParams; }
});
